### ISSUE TITLE
Add Hashicorp Vault as a remote state store

### ISSUE THREAD

**Issue by vaerh** [opentofu/opentofu#907]
### Summary

When using IaC, Hashicorp Vault is often used to properly store secrets. Given that the structure of Vault allows to work correctly with JSON objects, I would like to be able to use Vault as a remote backend for storing states.
In this case we have a role-based vault with versioning capability. The correct lifecycle of storing states, which may potentially contain sensitive data, is provided, and when using provider-framework the possibility of changing (e.g. hashing) Sensitive fields in the statefile is excluded. 

### Problem Statement

Ability to use Hashicorp Vault as a secret and state storage at the same time.

### User-facing description

![image](https://github.com/opentofu/opentofu/assets/64400271/499da175-afa1-438c-81b9-8c7e24175362)

### Technical Description

One backend will be added to the list of existing backends.

```go
// Init initializes the backends map with all our hardcoded backends.
func Init(services *disco.Disco) {
...
	backends = map[string]backend.InitFn{
...
		"s3":         func() backend.Backend { return backendS3.New() },
		"vault":      func() backend.Backend { return backendVault.New() },
``` 

### Rationale and alternatives

Adding or not adding new functionality will not change exactly anything, except for small desires of some individuals.

### Downsides

Possible disadvantages have been eliminated during development and testing, but want a competent review.

### Unresolved Questions

Since the addition of the backend was accomplished by reverse engineering the existing remote storage, it is necessary to verify the correctness of the interaction logic.

### Proof of Concept

#906 
``` shell
vault server -dev -dev-root-token-id=education
```

``` hcl
terraform {
  backend vault {
    address      = "http://127.0.0.1:8200"
    path         = "states"
    access_token = "education"
    gzip         = false
    lock         = true
  }
}
```

**Comment by kislerdm** [comment]
Hey @vaerh! Thank you for submitting the RFC! 

Could you please clarify the main motivation for using vault as the backend? Did I get it right that it's mostly because it allows to store the state with sensitive data securely?

If security is the main concern, could you please check the RFC about state encryption - #297, maybe it covers your problem. Thanks!

Another note, it's worth checking it the Vault license change could potentially affect OpenTofu users who would want to use hosted Vault as the backend.

**Comment by vaerh** [comment]
Yes #297 partially covers my request except for one point, we will still have to have additional shared storage deployed locally or an external data provider.
Thanks!

**Comment by cube2222** [comment]
Hey @vaerh, thanks for submitting this! Also, good point @kislerdm!

One note, to set timeline expectations, we're generally waiting until the stable release is out with any meaningful enhancements.

Now regarding the RFC itself. The main thing I'm interested in here is understanding community interest and how much people would use this. A feature like this will require a maintenance effort long-term, so we should always weigh that.

In this case, I'm wondering what fraction of people moving to OpenTofu actually want to keep using Vault, and get even more locked-in to it. Basically, if you're moving to OpenTofu, presumably you might want to migrate off of Vault too. However, I might be completely mistaken here. Curious of others' thoughts (esp. people who plan to use OpenTofu alongside Vault long-term!).

**Comment by vaerh** [comment]
Ok, let's wait for the community's opinion.

**Comment by iodeslykos** [comment]
I'd believe it likely that organizations utilizing Vault would be hard pressed to move, but if they were looking for an alternative there is still good reason to consider this feature now that OpenBao has been announced.

[OpenBao](https://github.com/openbao/openbao/tree/development)

**Comment by endocrimes** [comment]
The bigger issue here is that Terraform State can be _relatively_ large - and depending on the remote vault configuration - could cause denial of service issues across the cluster - especially when using Consul as the storage medium.

Consul can't chunk writes, so Consul will bounce the whole storage through Raft on every update - which in particularly large cases will cause Consul servers to start missing heartbeats and result in leadership instability.

This would be more foot-gun than feature.

**Comment by MatthewJohn** [comment]
Hey :)

Apologies, I accidentality started working on this - I got fed up and wondered what it would be like - I hadn't realised there was an issue already.
I took a look at the consul implementation (without thinking too much) and inherently included the functionality for chunking.
There's plenty still to do, as noted on the PR (https://github.com/opentofu/opentofu/pull/1680), but I published it for at least others to see :)

Matt

**Comment by vaerh** [comment]
Hi, I don't see a problem with an alternative implementation for our idea. The only thing that I didn't put into my implementation is the chunking, but I didn't test this implementation on large state files. 
In any case, I will be glad if we see this functional in OT  in some form.

**Comment by cipherboy** [comment]
One commentary I think I might add is that K/V with Raft might have a smaller file size limit than you'd expect. I think this can be solved by using multiple entries, but if the state file exceeds 1MB (the default upper limit in HashiCorp Vault and OpenBao), you might need to ask users to tweak the server configuration to [raise `max_storage_entry`](https://openbao.org/docs/configuration/storage/raft/#raft-parameters).

**Comment by MatthewJohn** [comment]
Hi @cipherboy , chunking can be used (as done in https://github.com/opentofu/opentofu/pull/1680) and is already done for the consul remote state.

**Comment by cipherboy** [comment]
@MatthewJohn Ah cool :-) How do you do the chunking transactionally? Would external transactions similar to etcd be useful here?

**Comment by roehrijn** [comment]
another opinion from the community: I would love to see this implemented. It would mitigate a bootstrapping problem which raises on every new project: In order to have a place to store state one must provision some state storage without support of the particular target state backend (e.g. you need to provision some S3 + permissions with local state backend to enable usage of s3 backend - this is not very nice). On the other hand we usually have a Vault instance (and hopefully OpenBao in the future) deployed as part of our CI/CD infrastructure. This makes it a perfect place to store state. 

<sub>Jan Roehrich <jan.roehrich@mercedes-benz.com>, Mercedes-Benz Tech Innovation GmbH, [legal info/Impressum](https://github.com/mercedes-benz/foss/blob/master/LEGAL_IMPRINT.md)</sub>

**Comment by apparentlymart** [comment]
A long, long time ago when I was first researching the topic of encrypting state storage (for https://github.com/hashicorp/terraform/issues/9556) I found https://github.com/hashicorp/terraform/issues/5374, which prompted me to ask for the HashiCorp Vault team's feedback on what they thought was an appropriate design.

Unfortunately I asked that question in a non-public place -- at the time I was an external community maintainer of Terraform, and so I was invited as a single-channel participant in HashiCorp's corporate Slack instance -- so I cannot link to the answer or directly quote it, but the essence I took from that feedback was: Storing a large artifact like a Terraform/OpenTofu state snapshot (whether as a single object or as multiple chunks) was not within the expected design parameters of Vault's generic key/value store and so they recommended _strongly_ against it, and suggested instead using Vault's "transit" engine to produce a ciphertext for storage elsewhere. That recommendation was, in large part, what motivated my proposal in https://github.com/hashicorp/terraform/issues/9556 [^1].

Since I originally asked that question in 2016, and thus Vault/OpenBao have likely changed a lot in the intervening eight years, and so that older answer might no longer be relevant. @cipherboy's questions above suggest to me that it might still be unwise to use OpenBao in this way, though. It would be unfortunate to add a feature that makes it seem like we're recommending this usage if there is risk of OpenBao losing data, suffering a performance degradation, or failing with an error when a state snapshot gets too large.

Of course, if we move ahead with "Backends as plugins" (https://github.com/opentofu/opentofu/issues/382) then it would be possible for a third-party to offer such an integration regardless of whether or not it's a good idea :grinning: but I'm still curious to learn if the advice I recieved in 2016 remains valid or if the situation has changed enough to reconsider.

---

@cipherboy: would you be willing to share a more direct opinion on whether you'd consider it a good idea to use OpenBao in the way this proposal suggests?

If you're willing/able, I'd love to hear a more up-to-date version of that advice I got in 2016, especially if you can add some more color about _why_ it might or might not be a good idea (since I no longer remember all those details from the earlier discussion).

[^1]: Although the details are different because Terraform's remote state design was very different prior to Terraform v0.9 when I wrote the proposal, OpenTofu's current feature for state encryption is similar and so something like that is already available in OpenTofu, albeit with [the OpenBao support](https://opentofu.org/docs/language/state/encryption/#openbao-experimental) still only experimental at the time I'm writing this. My original proposal was eventually closed for a different reason, but that reason is off-topic for _this_ issue and I won't comment on it further here.

**Comment by cipherboy** [comment]
\o hey @apparentlymart -- I had been aware of the same general advice you had been given. Doing some digging, I think this issue would be a blocker: https://github.com/etcd-io/bbolt/issues/694 or alternatively, implementing something that [calls `Compact(...)` periodically](https://github.com/etcd-io/bbolt/issues/517), which is apparently the approach taken by [etcd](https://github.com/etcd-io/etcd/blob/1472b61bdd4f217b9e2099e019316ddf72283edf/etcdctl/README.md?plain=1#L983-L991).

In particular, the issue is bbolt doesn't do online defrags automatically as part of the transaction and also writes all values as a contiguous block... So as the state file grows, it'll be moved if bbolt can't expand the current value in-place, which subsequently leaves allocated-but-free gaps in the bbolt file. If we subsequently make additional writes and deletes of other entries, we may end up continually growing a sparse bbolt file. 

I think this would be fairly common, as most people will have a lot of other workloads attached to a single OpenBao instance than _just_ OpenTofu state file holding, so I'd expect a lot of other smaller entries to be written & deleted which cause OpenTofu's state file to grow the allocated space when it is updated.

But, it is one of those things that is highly workload dependent.

---

re @roehrijn:

>  It would mitigate a bootstrapping problem which raises on every new project: In order to have a place to store state one must provision some state storage without support of the particular target state backend (e.g. you need to provision some S3 + permissions with local state backend to enable usage of s3 backend - this is not very nice).

I am also cautious of circular dependencies. Many people set up (the upstreams of) OpenBao with OpenTofu too, so having OpenBao store the state file could make certain operations harder (e.g., if the state file OpenTofu uses to manage the OpenBao cluster is subsequently encrypted using Transit or this mechanism, one could imagine inadvertent changes to cluster availability may prevent the state file from being persisted.)

**Comment by MatthewJohn** [comment]
> I am also cautious of circular dependencies. Many people set up (the upstreams of) OpenBao with OpenTofu too, so having OpenBao store the state file could make certain operations harder (e.g., if the state file OpenTofu uses to manage the OpenBao cluster is subsequently encrypted using Transit or this mechanism, one could imagine inadvertent changes to cluster availability may prevent the state file from being persisted.)

Whilst I agree this _could_ be an issue - I wouldn't avoid impementing a feature due to lack of consideration of implementation by the end-user.
It could be argued that the state bucket that is used for S3 is managed by Terraform state which is held in the same S3 bucket.
On the flip side, if you thing of the order of magnitude difference - One team may handle the deployment of an OpenBao cluster, whilst a multitude of other teams could be using OpenBao for the encryption of their state.
The advantages of using OpenBao (encryption and in my use-case the authentication mechanism with strong OIDC and claim-based policies) for the remaining X% (large percentage) of use-cases by far outweighs the small inconvenience of "how do I store the state to deploy OpenBau"

**Comment by roehrijn** [comment]
> I am also cautious of circular dependencies. Many people set up (the upstreams of) OpenBao with OpenTofu too, so having OpenBao store the state file could make certain operations harder (e.g., if the state file OpenTofu uses to manage the OpenBao cluster is subsequently encrypted using Transit or this mechanism, one could imagine inadvertent changes to cluster availability may prevent the state file from being persisted.)

@cipherboy, I fully agree, but at least if OpenBao backend was implemented it was up to the user to decide, which one of the two circular dependencies they want to resolve.
